### PR TITLE
Stop using detached threads. All threads must be joined.

### PR DESCRIPTION
**Pull request by vjpai**
Removing the use of detached threads means that we will have fewer paths to maintain as we convert core to idiomatic C++. This is a preparatory PR for the C++ificiation of gpr thread.

There was only 1 use of detached threads in all of `src` : for the poll-cv engine.
There were several uses in `test`, though.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                           FILE SIZE
 ++++++++++++++ GROWING                             ++++++++++++++
  +0.6%     +72 src/core/lib/iomgr/ev_poll_posix.cc     +72  +0.6%
      [NEW]     +72 cache_harvest_locked                    +72  [NEW]
      +6.6%     +34 run_poll                                +34  +6.6%
      +2.3%      +8 grpc_init_poll_cv_posix                  +8  +2.3%
      +1.8%      +6 [Unmapped]                               +6  +1.8%
      +2.4%      +5 global_cv_fd_table_shutdown              +5  +2.4%

 -------------- SHRINKING                           --------------
  -0.1%    -853 [None]                              -3.04Ki  -0.0%
      -0.1%    -821 [Unmapped]                          -3.04Ki  -0.0%
      -3.4%     -40 [None]                                    0  [ = ]
  [DEL]    -147 src/core/lib/gpr/thd.cc                -147  [DEL]
      [DEL]     -54 gpr_thd_options_default                 -54  [DEL]
      [DEL]     -44 [Unmapped]                              -44  [DEL]
      [DEL]     -22 gpr_thd_options_is_detached             -22  [DEL]
      [DEL]     -19 gpr_thd_options_is_joinable             -19  [DEL]
      [DEL]      -4 gpr_thd_options_set_detached             -4  [DEL]
      [DEL]      -4 gpr_thd_options_set_joinable             -4  [DEL]
  -8.1%     -80 src/core/lib/gpr/thd_posix.cc           -80  -8.1%
      [DEL]    -540 gpr_thd_new                            -540  [DEL]
  -3.2%     -80 src/core/lib/iomgr/timer_manager.cc     -80  -3.2%
     -27.5%     -71 start_timer_thread_and_unlock           -71 -27.5%
     -12.0%      -9 [Unmapped]                               -9 -12.0%

  -0.1% -1.06Ki TOTAL                               -3.27Ki  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: src/core/lib/iomgr/ev_poll_posix.cc [hunk 1]
```diff
@@ -1437,9 +1444,24 @@ static void cache_destroy_locked(poll_args* args) {
     poll_cache.free_pollers = args->next;
   }
 
+  // Now move this args to the dead poller list for later join
+  if (poll_cache.dead_pollers != nullptr) {
+    poll_cache.dead_pollers->prev = args;
+  }
+  args->prev = nullptr;
+  args->next = poll_cache.dead_pollers;
   gpr_free(args);
```
Comment (kpayson64) [inline src/core/lib/iomgr/ev_poll_posix.cc]: It looks like you were correct, we weren't doing refcounting for destroying ```args```, but I suspect we will have to start doing that to make these threads joinable.

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                           FILE SIZE
 ++++++++++++++ GROWING                             ++++++++++++++
  +0.6%     +72 src/core/lib/iomgr/ev_poll_posix.cc     +72  +0.6%
      [NEW]     +72 cache_harvest_locked                    +72  [NEW]
      +6.6%     +34 run_poll                                +34  +6.6%
      +2.3%      +8 grpc_init_poll_cv_posix                  +8  +2.3%
      +1.8%      +6 [Unmapped]                               +6  +1.8%
      +2.4%      +5 global_cv_fd_table_shutdown              +5  +2.4%

 -------------- SHRINKING                           --------------
  -0.1%    -853 [None]                              -3.04Ki  -0.0%
      -0.1%    -821 [Unmapped]                          -3.04Ki  -0.0%
      -3.4%     -40 [None]                                    0  [ = ]
  [DEL]    -147 src/core/lib/gpr/thd.cc                -147  [DEL]
      [DEL]     -54 gpr_thd_options_default                 -54  [DEL]
      [DEL]     -44 [Unmapped]                              -44  [DEL]
      [DEL]     -22 gpr_thd_options_is_detached             -22  [DEL]
      [DEL]     -19 gpr_thd_options_is_joinable             -19  [DEL]
      [DEL]      -4 gpr_thd_options_set_detached             -4  [DEL]
      [DEL]      -4 gpr_thd_options_set_joinable             -4  [DEL]
  -8.1%     -80 src/core/lib/gpr/thd_posix.cc           -80  -8.1%
      [DEL]    -540 gpr_thd_new                            -540  [DEL]
  -3.2%     -80 src/core/lib/iomgr/timer_manager.cc     -80  -3.2%
     -27.5%     -71 start_timer_thread_and_unlock           -71 -27.5%
     -12.0%      -9 [Unmapped]                               -9 -12.0%

  -0.1% -1.06Ki TOTAL                               -3.27Ki  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: src/core/lib/iomgr/ev_poll_posix.cc (hunk 1 above)
Comment (vjpai) [reply src/core/lib/iomgr/ev_poll_posix.cc]: So is there anything else that can own args once it hits the freelist? I guess I'm not seeing the shared ownership here though I need to look more carefully.

Comment (vjpai) [reply src/core/lib/iomgr/ev_poll_posix.cc]: Never mind, I see it. Ok, we'll have to do something about it.

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                           FILE SIZE
 ++++++++++++++ GROWING                             ++++++++++++++
  +0.6%     +72 src/core/lib/iomgr/ev_poll_posix.cc     +72  +0.6%
      [NEW]     +72 cache_harvest_locked                    +72  [NEW]
      +5.8%     +30 run_poll                                +30  +5.8%
      +3.0%     +10 [Unmapped]                              +10  +3.0%
      +2.3%      +8 grpc_init_poll_cv_posix                  +8  +2.3%
      +2.4%      +5 global_cv_fd_table_shutdown              +5  +2.4%

 -------------- SHRINKING                           --------------
  -0.1%    -885 [None]                              -3.00Ki  -0.0%
      -0.2%    -853 [Unmapped]                          -3.00Ki  -0.0%
      -3.4%     -40 [None]                                    0  [ = ]
  [DEL]    -147 src/core/lib/gpr/thd.cc                -147  [DEL]
      [DEL]     -54 gpr_thd_options_default                 -54  [DEL]
      [DEL]     -44 [Unmapped]                              -44  [DEL]
      [DEL]     -22 gpr_thd_options_is_detached             -22  [DEL]
      [DEL]     -19 gpr_thd_options_is_joinable             -19  [DEL]
      [DEL]      -4 gpr_thd_options_set_detached             -4  [DEL]
      [DEL]      -4 gpr_thd_options_set_joinable             -4  [DEL]
  -8.1%     -80 src/core/lib/gpr/thd_posix.cc           -80  -8.1%
      [DEL]    -540 gpr_thd_new                            -540  [DEL]
  -3.2%     -80 src/core/lib/iomgr/timer_manager.cc     -80  -3.2%
     -27.5%     -71 start_timer_thread_and_unlock           -71 -27.5%
     -12.0%      -9 [Unmapped]                               -9 -12.0%

  -0.1% -1.09Ki TOTAL                               -3.23Ki  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                           FILE SIZE
 ++++++++++++++ GROWING                             ++++++++++++++
  +0.6%     +72 src/core/lib/iomgr/ev_poll_posix.cc     +72  +0.6%
      [NEW]     +72 cache_harvest_locked                    +72  [NEW]
      +5.8%     +30 run_poll                                +30  +5.8%
      +3.0%     +10 [Unmapped]                              +10  +3.0%
      +2.3%      +8 grpc_init_poll_cv_posix                  +8  +2.3%
      +2.4%      +5 global_cv_fd_table_shutdown              +5  +2.4%

 -------------- SHRINKING                           --------------
  -0.1%    -885 [None]                              -3.00Ki  -0.0%
      -0.2%    -853 [Unmapped]                          -3.00Ki  -0.0%
      -3.4%     -40 [None]                                    0  [ = ]
  [DEL]    -147 src/core/lib/gpr/thd.cc                -147  [DEL]
      [DEL]     -54 gpr_thd_options_default                 -54  [DEL]
      [DEL]     -44 [Unmapped]                              -44  [DEL]
      [DEL]     -22 gpr_thd_options_is_detached             -22  [DEL]
      [DEL]     -19 gpr_thd_options_is_joinable             -19  [DEL]
      [DEL]      -4 gpr_thd_options_set_detached             -4  [DEL]
      [DEL]      -4 gpr_thd_options_set_joinable             -4  [DEL]
  -8.1%     -80 src/core/lib/gpr/thd_posix.cc           -80  -8.1%
      [DEL]    -540 gpr_thd_new                            -540  [DEL]
  -3.2%     -80 src/core/lib/iomgr/timer_manager.cc     -80  -3.2%
     -27.5%     -71 start_timer_thread_and_unlock           -71 -27.5%
     -12.0%      -9 [Unmapped]                               -9 -12.0%

  -0.1% -1.09Ki TOTAL                               -3.23Ki  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                           FILE SIZE
 ++++++++++++++ GROWING                             ++++++++++++++
  +0.8%    +104 src/core/lib/iomgr/ev_poll_posix.cc    +104  +0.8%
      [NEW]     +72 cache_harvest_locked                    +72  [NEW]
      +8.9%     +46 run_poll                                +46  +8.9%
      +5.5%     +18 [Unmapped]                              +18  +5.5%
      +2.3%      +8 grpc_init_poll_cv_posix                  +8  +2.3%
      +2.4%      +5 global_cv_fd_table_shutdown              +5  +2.4%

 -------------- SHRINKING                           --------------
  -0.1%    -885 [None]                              -2.97Ki  -0.0%
      -0.2%    -853 [Unmapped]                          -2.97Ki  -0.0%
      -3.4%     -40 [None]                                    0  [ = ]
  [DEL]    -147 src/core/lib/gpr/thd.cc                -147  [DEL]
      [DEL]     -54 gpr_thd_options_default                 -54  [DEL]
      [DEL]     -44 [Unmapped]                              -44  [DEL]
      [DEL]     -22 gpr_thd_options_is_detached             -22  [DEL]
      [DEL]     -19 gpr_thd_options_is_joinable             -19  [DEL]
      [DEL]      -4 gpr_thd_options_set_detached             -4  [DEL]
      [DEL]      -4 gpr_thd_options_set_joinable             -4  [DEL]
  -8.1%     -80 src/core/lib/gpr/thd_posix.cc           -80  -8.1%
      [DEL]    -540 gpr_thd_new                            -540  [DEL]
  -3.2%     -80 src/core/lib/iomgr/timer_manager.cc     -80  -3.2%
     -27.5%     -71 start_timer_thread_and_unlock           -71 -27.5%
     -12.0%      -9 [Unmapped]                               -9 -12.0%

  -0.1% -1.06Ki TOTAL                               -3.17Ki  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by vjpai** [comment]
The TSAN thread leak thing is real and it's a problem. The source is that right now threads are only being harvested at shutdown, but that application calls exit which doesn't actually trigger a shutdown. I thought about using `atexit` but it's complicated because there's no guarantee of if `atexit` handlers are called before or after the thread really dies. I'm looking to see if there's another option.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                         FILE SIZE
 ++++++++++++++ GROWING                                           ++++++++++++++
  +2.5%    +312 src/core/lib/iomgr/ev_poll_posix.cc                  +312  +2.5%
       +27%    +137 run_poll                                             +137   +27%
      [NEW]    +123 cache_harvest_locked                                 +123  [NEW]
       +17%     +24 init_result(poll_args*) [clone .constprop.16]         +24   +17%
       +17%     +15 cache_insert_locked                                   +15   +17%
      +0.6%      +9 cvfd_poll                                              +9  +0.6%
      +2.3%      +8 grpc_init_poll_cv_posix                                +8  +2.3%
      +2.4%      +5 global_cv_fd_table_shutdown                            +5  +2.4%
      +1.5%      +3 cache_poller_locked                                    +3  +1.5%

 -------------- SHRINKING                                         --------------
  -0.1%    -861 [None]                                            -2.93Ki  -0.0%
      -0.2%    -829 [Unmapped]                                        -2.93Ki  -0.0%
      -3.4%     -40 [None]                                                  0  [ = ]
  [DEL]    -147 src/core/lib/gpr/thd.cc                              -147  [DEL]
      [DEL]     -54 gpr_thd_options_default                               -54  [DEL]
      [DEL]     -44 [Unmapped]                                            -44  [DEL]
      [DEL]     -22 gpr_thd_options_is_detached                           -22  [DEL]
      [DEL]     -19 gpr_thd_options_is_joinable                           -19  [DEL]
      [DEL]      -4 gpr_thd_options_set_detached                           -4  [DEL]
      [DEL]      -4 gpr_thd_options_set_joinable                           -4  [DEL]
  -8.1%     -80 src/core/lib/gpr/thd_posix.cc                         -80  -8.1%
      [DEL]    -540 gpr_thd_new                                          -540  [DEL]
  -3.2%     -80 src/core/lib/iomgr/timer_manager.cc                   -80  -3.2%
     -27.5%     -71 start_timer_thread_and_unlock                         -71 -27.5%
     -12.0%      -9 [Unmapped]                                             -9 -12.0%

  -0.1%    -856 TOTAL                                             -2.92Ki  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                         FILE SIZE
 ++++++++++++++ GROWING                                           ++++++++++++++
  +2.5%    +312 src/core/lib/iomgr/ev_poll_posix.cc                  +312  +2.5%
       +27%    +137 run_poll                                             +137   +27%
      [NEW]    +123 cache_harvest_locked                                 +123  [NEW]
       +17%     +24 init_result(poll_args*) [clone .constprop.16]         +24   +17%
       +17%     +15 cache_insert_locked                                   +15   +17%
      +0.6%      +9 cvfd_poll                                              +9  +0.6%
      +2.3%      +8 grpc_init_poll_cv_posix                                +8  +2.3%
      +2.4%      +5 global_cv_fd_table_shutdown                            +5  +2.4%
      +1.5%      +3 cache_poller_locked                                    +3  +1.5%

 -------------- SHRINKING                                         --------------
  -0.1%    -861 [None]                                            -2.93Ki  -0.0%
      -0.2%    -829 [Unmapped]                                        -2.93Ki  -0.0%
      -3.4%     -40 [None]                                                  0  [ = ]
  [DEL]    -147 src/core/lib/gpr/thd.cc                              -147  [DEL]
      [DEL]     -54 gpr_thd_options_default                               -54  [DEL]
      [DEL]     -44 [Unmapped]                                            -44  [DEL]
      [DEL]     -22 gpr_thd_options_is_detached                           -22  [DEL]
      [DEL]     -19 gpr_thd_options_is_joinable                           -19  [DEL]
      [DEL]      -4 gpr_thd_options_set_detached                           -4  [DEL]
      [DEL]      -4 gpr_thd_options_set_joinable                           -4  [DEL]
  -8.1%     -80 src/core/lib/gpr/thd_posix.cc                         -80  -8.1%
      [DEL]    -540 gpr_thd_new                                          -540  [DEL]
  -3.2%     -80 src/core/lib/iomgr/timer_manager.cc                   -80  -3.2%
     -27.5%     -71 start_timer_thread_and_unlock                         -71 -27.5%
     -12.0%      -9 [Unmapped]                                             -9 -12.0%

  -0.1%    -856 TOTAL                                             -2.92Ki  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by vjpai** [comment]
@kpayson64 : The latest commit (5e1298f) does some 2-phase synchronization to join poll-cv threads cleanly without any thread leaks on early exit.

The remaining failures are unrelated.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                         FILE SIZE
 ++++++++++++++ GROWING                                           ++++++++++++++
  +2.5%    +312 src/core/lib/iomgr/ev_poll_posix.cc                  +312  +2.5%
       +27%    +137 run_poll                                             +137   +27%
      [NEW]    +123 cache_harvest_locked                                 +123  [NEW]
       +17%     +24 init_result(poll_args*) [clone .constprop.16]         +24   +17%
       +17%     +15 cache_insert_locked                                   +15   +17%
      +0.6%      +9 cvfd_poll                                              +9  +0.6%
      +2.3%      +8 grpc_init_poll_cv_posix                                +8  +2.3%
      +2.4%      +5 global_cv_fd_table_shutdown                            +5  +2.4%
      +1.5%      +3 cache_poller_locked                                    +3  +1.5%

 -------------- SHRINKING                                         --------------
  -0.1%    -861 [None]                                            -2.93Ki  -0.0%
      -0.2%    -829 [Unmapped]                                        -2.93Ki  -0.0%
      -3.4%     -40 [None]                                                  0  [ = ]
  [DEL]    -147 src/core/lib/gpr/thd.cc                              -147  [DEL]
      [DEL]     -54 gpr_thd_options_default                               -54  [DEL]
      [DEL]     -44 [Unmapped]                                            -44  [DEL]
      [DEL]     -22 gpr_thd_options_is_detached                           -22  [DEL]
      [DEL]     -19 gpr_thd_options_is_joinable                           -19  [DEL]
      [DEL]      -4 gpr_thd_options_set_detached                           -4  [DEL]
      [DEL]      -4 gpr_thd_options_set_joinable                           -4  [DEL]
  -8.1%     -80 src/core/lib/gpr/thd_posix.cc                         -80  -8.1%
      [DEL]    -540 gpr_thd_new                                          -540  [DEL]
  -3.2%     -80 src/core/lib/iomgr/timer_manager.cc                   -80  -3.2%
     -27.5%     -71 start_timer_thread_and_unlock                         -71 -27.5%
     -12.0%      -9 [Unmapped]                                             -9 -12.0%

  -0.1%    -856 TOTAL                                             -2.92Ki  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by vjpai** [comment]
Not only is the PR green, I also ran tsan tests of the previous thread leaks 500x on my desktop and got no failures, even with multiple antagonists. I think this works.

**Comment by vjpai** [comment]
This PR is subsumed by #14459

**Pull request merged**